### ISSUE TITLE
no-duplicate-dollar-variables doesn't respect scopes

### ISSUE THREAD

**Issue by robertjk** [kristerkari/stylelint-scss#277]
`no-duplicate-dollar-variables` rule reports variables with the same name but declared in different scopes. The snippet below produces error `Unexpected duplicate dollar variable $padding-x   scss/no-duplicate-dollar-variables`.

```
.FormControl-label {
  $padding-x: $padding-small-x;

  @include input-overlay;

  position: relative;
  top: 0.40em;
  margin-left: $margin-medium-x;
  padding-left: $padding-x;
  padding-right: $padding-x;
  font-size: $font-size-small;
}


.FormControl-input {
  $padding-x: $padding-medium-x;

  width: 100%;
  border: 1px solid $color-grey-light;
  border-radius: 0;
  padding-left: $padding-x;
  padding-right: $padding-x;
  height: $row-height;
}
```

For me it's pretty obvious that this shouldn't be reported. Am I wrong?

**Comment by kristerkari** [comment]
Thanks! The way the rule was created is that it looks for variables in the whole stylesheet's scope. I think that the rule could be changed to be a bit more clever to not warn for your use case.

So, this would warn (overwriting a variable from outer scope):

```scss
$a: 1;

.foo {
  $a: 2;
} 

.bar {
  $a: 3;
}
```

...but this would not warn (the variable is not defined in the outer scope):

```scss
.foo {
  $a: 2;
} 

.bar {
  $a: 3;
}
```

**Comment by chrismbarr** [comment]
Another false positive example I ran across today:

```scss
@if ($value == 'foo') {
    $my-value: 'bar';
} @else {
    $my-value: $value;
//  ^^^^^^^^^ Unexpected duplicate dollar variable $my-value (scss/no-duplicate-dollar-variables)  
}
```

**Comment by kristerkari** [comment]
@chrismbarr those are not actually false positives, to avoid that you can use the ignore options of the rule:  
https://github.com/kristerkari/stylelint-scss/blob/master/src/rules/no-duplicate-dollar-variables/README.md#ignoreinsideatrules-array-of-at-rules

**Comment by kristerkari** [comment]
This is now fixed in v3.15.0
https://github.com/kristerkari/stylelint-scss/releases/tag/3.15.0

Thanks everyone for helping out!

**Comment by masi** [comment]
WIth 3.18 the rule triggers:

```
.foobar {
  $gapWidth: 5px;

  @media (min-width: 600px) {
    $gapWidth: 15px;
  }
}

```

I had assumed that the media query defines a scope.
